test/2-nodes.js

```diff
@@ -0,0 +1,544 @@
+/* eslint-env mocha */
+/* eslint max-nested-callbacks: ["error", 5] */
+'use strict'
+
+const chai = require('chai')
+chai.use(require('dirty-chai'))
+chai.use(require('chai-spies'))
+const expect = chai.expect
+const parallel = require('async/parallel')
+const series = require('async/series')
+const times = require('lodash/times')
+const promisify = require('promisify-es6')
+
+const GossipSub = require('../src')
+const utils = require('./utils')
+const first = utils.first
+const createNode = utils.createNode
+const expectSet = utils.expectSet
+
+describe('basics between 2 nodes', () => {
+  describe('fresh nodes', () => {
```

Comment:
We should be able to run every single test in this suite of tests individually, which is not the case because we have "tests" for:

- `mount the pubsub protocol`
-  `start both GossipSubs`
- `Dial from nodeA to nodeB`

Those tests are essential for the remaining ones. For instance, if I do a `it.only(...)` for one of the following tests, it will fail because it will need the previous ones to run.

What I recommend here is to have multiple befores for all the code that must run for each test:

```js
// Create nodes
before(async () => 
....

// mount the pubsub protocol
before(....

// start both GossipSubs
before(...

// Dial from nodeA to nodeB
before(...
```

However, there are other dependencies here between the remaining tests. In this case, you can also use `beforeEach` and `afterEach`, which will run before and after each test, in order to setup specific behavior for each of them.

There are other approaches as well, I am ok with other solutions but the tests must be completely isolated. If you need assistance on this, just ping me on IRC

Reply:
Yeah, these tests are not the best. They are inherited from `js-libp2p-floodsub`, and that repo has the same issues you mention here.

I'm thinking it may almost be easier to scrap the existing tests and redo them from scratch, testing and setting up each item independently.
They also don't test the gossipsub components at all (I noticed several bugs in the rpc handling while addressing your review).

I'll work on revamping the testing to be more complete and flexible.

Reply:
yes! `libp2p-floodsub` tests are terrible and definitely not a good example! Unfortunately, I got the responsibility for the repo in that state, and I did not have the bandwidth to refactor them.